// mano-legacy/hash-nav.js
// Switches 'active' class on elements with id's matching url hash

'use strict';

var $      = require('./main')
  , elMain = require('./element#/event')

  , currentTarget = null, updateTo;

require('./element#/class');
require('./element#/mano-event');

exports.updateTo = updateTo = function (hash) {
	var nuTarget;
	if (!hash) hash = null;
	else hash = String(hash);
	nuTarget = hash && $(hash);
	if (nuTarget !== currentTarget) {
		if (currentTarget) {
			currentTarget.removeClass('active');
			currentTarget.addClass('inactive');
		}
		currentTarget = nuTarget;
		if (currentTarget) {
			currentTarget.removeClass('inactive');
			currentTarget.addClass('active');
		}
		exports.emit('update', currentTarget);
	}
};

exports.update = function () {
	var hash = location.hash;
	exports.emit('update:before');
	if (hash) hash = hash.slice(1);
	updateTo(hash);
};

exports.on = elMain.on;
exports.once = elMain.once;
exports.off = elMain.off;
exports.emit = elMain.emit;

elMain.addEvent.call(window, 'hashchange', exports.update);
elMain.addEvent.call(document, 'click', function () { setTimeout(exports.update, 0); });
require('./on-document-ready')(exports.update);
elMain.once.call(document, 'mano:load', exports.update);

exports.update();
exports.getCurrentTarget = function () { return currentTarget; };
