// mano-legacy/hash-nav-ordered-list-controls.js
// Controls handling for hash nav ordered list

'use strict';

var once    = require('timers-ext/once')
  , $       = require('./main')
  , hashNav = require('./hash-nav')

  , min = Math.min, max = Math.max;

require('./element#/class');

var update = function () {
	var index = 1, l = this.list.length;
	if (location.hash.indexOf('#' + this.prefix) === 0) {
		index = min(max(Number(location.hash.slice(this.prefix.length + 2)) || 1, 1), l);
	}

	// Previous control
	if ((index <= 1) || (l <= 1)) {
		this.controlPrev.removeAttribute('href');
		this.controlPrev.addClass('disabled');
	} else {
		this.controlPrev.removeClass('disabled');
		if (index > 2) this.controlPrev.setAttribute('href', '#' + this.prefix + '-' + (index - 1));
		else this.controlPrev.setAttribute('href', location.pathname + location.search);
	}

	// Next control
	if (index >= l) {
		this.controlNext.removeAttribute('href');
		this.controlNext.addClass('disabled');
	} else {
		this.controlNext.setAttribute('href', '#' + this.prefix + '-' + (index + 1));
		this.controlNext.removeClass('disabled');
	}

	// Status
	if (this.statusCurrent) this.statusCurrent.data = index;
};

var Nav = function (nav, list, prefix) {
	var i, l, els, el, self = this;
	nav = $(nav);
	els = nav.getElementsByTagName('a');
	i = 0;
	l = els.length;
	for (i = 0; i < l; ++i) {
		el = $(els[i]);
		if (el.hasClass('previous')) this.controlPrev = el;
		else if (el.hasClass('next')) this.controlNext = el;
	}
	els = nav.getElementsByTagName('span');
	i = 0;
	l = els.length;
	for (i = 0; i < l; ++i) {
		el = $(els[i]);
		if (el.hasClass('current-index')) {
			this.statusCurrent = el.firstChild;
			if (!this.statusCurrent) this.statusCurrent = el.appendChild(document.createTextNode());
			break;
		}
	}
	this.list = $(list).getElementsByTagName('li');
	this.prefix = prefix;
	this.update = once(function () { update.call(self); });
};

module.exports = $.hashNavOrderedListControls = function (dom, ul, idPrefix) {
	var nav = new Nav(dom, ul, idPrefix);
	hashNav.on('update:before', nav.update);
	if (typeof document.on === 'function') document.on('dbupdate', nav.update);
	nav.update();
	return nav;
};
