// location-emitter/test/client.js
'use strict';

var forEach = require('es5-ext/object/for-each');

module.exports = function (t, a) {
	var window, location, le, invoked = 0;
	location = {
		hash: '#hash',
		host: 'example.com:8080',
		hostname: 'example.com',
		href: 'http://example.com:8080/path/?search#hash',
		origin: 'http://example.com:8080',
		pathname: '/path/',
		port: '8080',
		protocol: 'http:',
		search: '?search'
	};
	window = {
		history: {
			pushState: function () {
				location.hash = '#bar';
				location.href = 'http://example.com:8080/other/?foo#bar';
				location.pathname = '/other/';
				location.search = '?foo';
			}
		},
		location: location,
		addEventListener: function () {}
	};

	le = t(window);

	forEach(location, function (value, name) {
		a(le[name], value, name);
	});

	le.on('change', function (url, search, hash) {
		a(url, '/other/', "Event: url");
		a(search, '?foo', "Event: search");
		a(hash, '#bar', "Event: hash");
		++invoked;
	});
	le.on('/other/', function (search, hash) {
		a(search, '?foo', "Event: search");
		a(hash, '#bar', "Event: hash");
		++invoked;
	});
	le.goto('/other/?foo#bar');
	a(invoked, 2, "Events invoked");

	forEach(location, function (value, name) {
		a(le[name], value, "After update: " + name);
	});
};
